### PR TITLE
datapack validation (GSI-468)

### PR DESCRIPTION
**Pull request by KerstenBreuer**
```
- moved schemapack models into spec subpackage
- added models for loading datapacks
- translate Pydantic validation errors for both schemapack and
   datapack models into custom format errors upon loading
- added examples and tests for valid and invalid datasets
- added validation of datapacks against schemapacks based on a plugin system
- updated the template
```

Please note, the most relevant changes are the datapack models in ./src/schemapack/spec/datapack.py as well as the validation logic in ./src/schemapack/validation. The rest is mostly re-arrangements of existing code. The validation plugins all follow a common pattern. While there are quite a few of them, each of them is simple. Since this is a POC, feel free to skip some details as long as the main skeleton is reviewed.

### CODE REVIEW

File: tests/test_main.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    VALID_SCHEMAPACK_PATHS,
+)
+
+
+@pytest.mark.parametrize(
+    "name, path", VALID_DATAPACK_PATHS.items(), ids=VALID_DATAPACK_PATHS
+)
+def test_load_and_validate_valid(name: str, path: Path):
+    """Test load_and_validate function with valid schemapack and valid datapacks."""
+    schemapack_name, error_type, *_ = tuple(name.split("."))
```
Comment (Cito) [inline tests/test_main.py:37]:
```suggestion
    schemapack_name = name.split(".", 1)[0]
```

File: tests/test_main.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+    _ = load_and_validate(schemapack_path=schemapack_path, datapack_path=path)
+
+
+@pytest.mark.parametrize(
+    "name, path", INVALID_DATAPACK_PATHS.items(), ids=INVALID_DATAPACK_PATHS
+)
+def test_load_and_validate_invalid(name: str, path: Path):
+    """Test load_and_validate function with valid schemapack but invalid datapacks."""
+    schemapack_name, error_type, *_ = tuple(name.split("."))
```
[the remark this answers is not in the document]
Comment (Cito) [reply tests/test_main.py:48]: ok, not much simpler...

File: src/schemapack/validation/main.py [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+
+        Args:
+            schemapack:
+                The schemapack to validate against.
+            add_global_plugins:
+                Global validation plugins to use in addition to the default onces.
+            add_class_plugins:
+                Class validation plugins to use in addition to the default onces.
+            add_resource_plugins:
+                Resource validation plugins to use in addition to the default onces.
```
Comment (Cito) [inline src/schemapack/validation/main.py:193]:
```suggestion
            add_global_plugins:
                Global validation plugins to use in addition to the default ones.
            add_class_plugins:
                Class validation plugins to use in addition to the default ones.
            add_resource_plugins:
                Resource validation plugins to use in addition to the default ones.
```

File: src/schemapack/spec/datapack.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        ),
+    )
+
+    resources: dict[ClassName, dict[ResourceId, Resource]] = Field(
+        ...,
+        description=(
+            "A nested dictionary containing resources per class name (keys on the first"
+            + " level) and resource ID (keys on the second level). Each class defined"
+            + " in the schemapack must be present even if no resources are defined for"
+            + " for it in this datapack."
```
Comment (Cito) [inline src/schemapack/spec/datapack.py:82]:
```suggestion
            + " it in this datapack."
```

File: src/schemapack/spec/datapack.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        ),
+    )
+
+    root: Optional[ResourceId] = Field(
+        None,
+        description=(
+            "The ID of the root resource which must be of the class defined as root"
+            + " in the schemapack. This must and may be only specified if the schempack"
```
Comment (Cito) [inline src/schemapack/spec/datapack.py:90]:
```suggestion
            + " in the schemapack. This must and may be only specified if the schemapack"
```

**Review by Cito — APPROVED** [review]
Looks good, and tests still run very fast. Just some cosmetic changes and suggestions.

File: src/schemapack/validation/base.py [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+
+        Raises:
+            schemapack.exceptions.DataValidationError: If validation fails.
+        """
+        raise NotImplementedError
+
+
+class ClassValidationPlugin(ABC):
+    """Abstract class for defining class-scoped validation plugin. Implementing objects
+    perform validation with respect to one specific schemapack class.
```
[the remark this answers is not in the document]
Comment (KerstenBreuer) [reply src/schemapack/validation/base.py:64]: I changed it to "Instances of implemeting concrete classes" as this really means the instance not the classes.

**Review by Cito — APPROVED** [review]

**Pull request merged**